### PR TITLE
API Integration

### PR DESCRIPTION
**Pull request by gzamudio**
## Summary
- Added redux on tic tac toe
- Added match history
- Added spinner

# Trello Card
- https://trello.com/c/Uadv12DA

### CODE REVIEW

File: react/tic-tac-toe/src/redux/match/reducer.js [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  xIsNext: true,
+  matches: [],
+  isLoading: false
+};
+
+function reducer(state = initialState, action) {
+  switch (action.type) {
+    case actions.SET_BOARD: {
+      const { history, stepNumber, xIsNext } = state;
+      const newHistory = history.slice(0, stepNumber + 1);
```
Comment (Dorian30) [inline react/tic-tac-toe/src/redux/match/reducer.js]: I think there is no need to make a newHistory array 👀You are copying the current history, but not adding or removing anything else. You can do the same steps with just the current history.

```js
const squares = [...history[history.length - 1].squares];
if (getWinner(squares) || squares[action.payload]) return state; 
squares[action.payload] = xIsNext ? 'X' : 'O';
return { ...state, history: [...history, { squares }], stepNumber: stepNumber + 1, xIsNext: !xIsNext };
```

File: react/tic-tac-toe/src/app/screens/Game/index.js [hunk 2]
```diff
@@ -40,26 +32,56 @@ class Game extends Component {
   });
 
   render() {
-    const { history, stepNumber, xIsNext } = this.state;
-    const current = [...history[stepNumber]];
-    const winner = getWinner(current.squares);
+    const { history, stepNumber, xIsNext, matchHistory, isLoading } = this.props;
+    const squares = history[stepNumber].squares.slice();
```
Comment (Dorian30) [inline react/tic-tac-toe/src/app/screens/Game/index.js]: Use the spread operator for copying an array.

```suggestion
    const squares = [...history[stepNumber].squares];
```

File: react/tic-tac-toe/src/app/screens/Game/index.js [hunk 3]
```diff
@@ -40,26 +32,56 @@ class Game extends Component {
...
+    const winner = getWinner(squares);
     const moves = this.getMoves(history);
     const status = winner ? `Winner: ${winner}` : `Next player: ${xIsNext ? 'X' : 'O'}`;
```
Comment (Dorian30) [inline react/tic-tac-toe/src/app/screens/Game/index.js]: No need to change this, but we usually store strings in constant variables.

File: react/tic-tac-toe/src/app/screens/Game/components/MatchHistory/index.js [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+import { matchPropType } from '../../../../../constants/propTypes';
+
+import styles from './styles.module.scss';
+
+class MatchHistory extends Component {
+  renderMatches() {
+    return this.props.matches.map((match, index) => {
+      const playerOneClass = match.winner === 'tie' ? styles.tie : match.winner === 'player_one' ? styles.winner : styles.loser;
+      const playerTwoClass = match.winner === 'tie' ? styles.tie : match.winner === 'player_one' ? styles.loser : styles.winner;
```
Comment (Dorian30) [inline react/tic-tac-toe/src/app/screens/Game/components/MatchHistory/index.js]: Hold these strings in constant variables like so 

```js
export const PLAYER_ONE: 'player_one';
export const PLAYER_TWO: 'player_two';
export const TIE: 'tie'; 
```

File: react/tic-tac-toe/src/redux/match/reducer.js (hunk 1 above)
Comment (gzamudio) [reply react/tic-tac-toe/src/redux/match/reducer.js]: mm but i dont need the last element from the history array, i need the element at the 'stepNumber' position.

File: react/tic-tac-toe/src/scss/_colors.scss [hunk 5]
```diff
@@ -1,2 +1,6 @@
-$white: #FFF;
+$black: #000;
 $dust-gray: #999;
+$green-04: rgba(0, 255, 0, 0.4);
+$red-04: rgba(255, 0, 0, 0.4);
+$white: #FFF;
+$yellow-04: rgba(255, 255, 0, 0.4);
```
Comment (Dorian30) [inline react/tic-tac-toe/src/scss/_colors.scss]: Missing white line at the end of this file 👀

**Pull request merged**